File: CPP/src/gui/puzzlesolver/puzzlesolver.cc
# include "puzzlesolver.hh"

Gui::PuzzleSolver::PuzzleSolver ( QWidget* parent )
   : QWidget ( parent )
{
   field = 0;
   buildGui ();
}

Gui::PuzzleSolver::~PuzzleSolver ( void )
{
}

void Gui::PuzzleSolver::buildGui ( void )
{
   horizontal_layout = new QHBoxLayout ( this );
   this->setLayout ( horizontal_layout );
   
   vertical_layout_left = new QVBoxLayout ();
   vertical_layout_right = new QVBoxLayout ();
   
   horizontal_layout->addLayout ( vertical_layout_left );
   horizontal_layout->addLayout ( vertical_layout_right );
   
   viewer = new Gui::PuzzleViewer ();
   actions = new Gui::PuzzleActions ();
   
   vertical_layout_left->addWidget ( viewer );
   
   vertical_layout_right->addStretch ();
   vertical_layout_right->addWidget ( actions );
   vertical_layout_right->addStretch ();
   
   connect ( actions , SIGNAL ( openFile () ) , this , SLOT ( openFile () ) );
   connect ( actions , SIGNAL ( runPuzzle () ) , this , SLOT ( runPuzzle () ) );
   
   return;
}

void Gui::PuzzleSolver::openFile ( void )
{
   qDebug () << "Gui::PuzzleSolver::openFile: Opening puzzle file";
   
   std::ifstream file;
   
   QString file_path;
   file_path = QFileDialog::getOpenFileName ( this ,
                                              "Open File" ,
                                              "" ,
                                              "Files (*)"
                                            );
   
   qDebug () << "Puzzle file path: " << file_path;
   
   file.open ( file_path.toStdString ().c_str () , std::ios::in );
   
   if ( !file.is_open () )
   {
      qDebug () << "Can't open file";
      return;
   }
   
   // Get puzzle size
   int puzzle_width , puzzle_height;
   file >> puzzle_width >> puzzle_height;
   
   if ( field != 0 )
   {
      delete field;
   }
   
   field = new Core::Field ( puzzle_width , puzzle_height );
   
   int piece_amount;
   file >> piece_amount;
   puzzle_pieces.clear ();
   
   for ( int i = 0; i < piece_amount; i++ )
   {
      Core::Piece piece;
      piece.load ( file );
      puzzle_pieces.push_back ( piece );
   }
   
   viewer->clean ();
   viewer->setSize ( puzzle_width , puzzle_height );
   
   qDebug () << "Puzzle file loaded: " << puzzle_width << "x" << puzzle_height;
   
   return;
}

void Gui::PuzzleSolver::runPuzzle ( void )
{
   if ( field == 0 )
   {
      return;
   }
   
   viewer->clean ();
   
   qDebug () << "Gui::PuzzleSolver::runPuzzle: Running solver";
   
   if ( !solve ( (*field) , puzzle_pieces , 0 ) )
   {
      qDebug () << "No solution found" << endl;
      QMessageBox::critical ( this , "Search complete" , "No solution found" );
   }
   else
   {
      QMessageBox::information ( this , "Search complete" , "Solution found" );
   }
   
   return;
}

bool Gui::PuzzleSolver::solve ( Core::Field field , vector< Core::Piece >& working_pieces , int current_piece )
{
   qDebug () << "Adding piece ID " << current_piece << endl << endl;
   
   if ( current_piece == (int)(working_pieces.size ()) )
   {
      qDebug () << "Solution found:" << endl << endl;
      field.print ();
      viewer->update ( field );
      
      return ( true );
   }
   
   working_pieces[ current_piece ].print ();
   
   qDebug () << endl << "Into this field:" << endl;
   field.print ();
   viewer->update ( field );
   qDebug () << endl;
   
   bool solved = false;
   
   qDebug () << "Trying to place..." << endl;
   
   for ( int y = 0; y < field.getHeight () && !solved; y++ )
   {
      for ( int x = 0; x < field.getWidth () && !solved; x++ )
      {
         if ( field.canBeInserted ( working_pieces[ current_piece ] , x , y ) )
         {
            Core::Field temp_field = field;
            temp_field.insertPiece ( working_pieces[ current_piece ] , x , y );
            
            qDebug () << endl <<"Piece added at " << x << "," << y << endl;
            
            qDebug () << "New field: " << endl;
            temp_field.print ();
            viewer->update ( temp_field );
            
            qDebug () << endl;
            
            solved = solve ( temp_field , working_pieces , current_piece + 1 );
            
            if ( !solved )
            {
               qDebug () << "Trying another position with piece " << current_piece << endl;
            }
         }
      }
   }
   
   return ( solved );
}

File: Python/gui/PuzzleSolver.py
#!/usr/bin/env python3

from PyQt4 import QtGui
from collections import deque
import copy

from .PuzzleViewer import PuzzleViewer
from .PuzzleActions import PuzzleActions
from core.Field import Field
from core.Piece import Piece

class PuzzleSolver ( QtGui.QWidget ):
   """Main container of the program"""
   def __init__ ( self ):
      super ( PuzzleSolver , self ).__init__ ()
      # Create the main container
      self.__horizontal_layout = QtGui.QHBoxLayout ()
      # Create components
      self.__puzzle_actions = PuzzleActions ()
      self.__puzzle_viewer = PuzzleViewer ()
      # Add components
      self.setLayout ( self.__horizontal_layout )
      self.__horizontal_layout.addWidget ( self.__puzzle_viewer )
      self.__horizontal_layout.addWidget ( self.__puzzle_actions )
      # Connect signals and slots
      self.__puzzle_actions.signal_open_file.connect ( self.__openFile )
      self.__puzzle_actions.signal_run_puzzle.connect ( self.__runPuzzle )
      self.__puzzle_pieces = []
      
   def __openFile ( self ):
      print ( "PuzzleSolver::openFile: Opening puzzle file" )
      file_path = QtGui.QFileDialog ().getOpenFileName ( self , "Open file" , "" , "Files (*)" )
      print ( "Trying to open file:" , file_path )
      
      try:
         file = open ( file_path , "r" )
      except IOError:
         print ( "Error opening file. Try again" )
         return
      
      file_contents = []
      
      for line in file:
         clean_line = line.rstrip ()
         words = clean_line.split ( " " )
         if len ( words ) == 1:
            if words[ 0 ] == "":
               continue
         file_contents.extend ( words )
         
      # Remove empty words
      while file_contents.count ( "" ) > 0:
         file_contents.remove ( "" )
      
      file.close ()
      # Convert the contents of the file into a queue
      file_queue = deque ( file_contents )
      
      puzzle_width = int ( file_queue.popleft () )
      puzzle_height = int ( file_queue.popleft () )
      self.__field = Field ( puzzle_width , puzzle_height )
      
      piece_amount = int ( file_queue.popleft () )
      
      # Clean all
      self.__puzzle_pieces[:] = []

      for i in range ( 0 , piece_amount ):
         piece = Piece ()
         piece.load ( file_queue )
         self.__puzzle_pieces.append ( piece )
         
      self.__puzzle_viewer.clean ()
      self.__puzzle_viewer.setSize ( puzzle_width , puzzle_height )
      
      print ( "File loaded" )
      
   def __runPuzzle ( self ):
      print ( "PuzzleSolver::runPuzzle: Running puzzle solver" )
      
      self.__puzzle_viewer.clean ()
      
      field_copy = copy.deepcopy ( self.__field )
      
      print ( "Original field size:" , self.__field.getWidth (), "x", self.__field.getHeight () )
      print ( "Copy field size:" , field_copy.getWidth (), "x", field_copy.getHeight () )
      
      if not self.__solve ( field_copy , self.__puzzle_pieces , 0 ):
         print ( "Solution not found" )
         QtGui.QMessageBox ().critical ( self , "Search complete" , "No solution found" )
      else:
         QtGui.QMessageBox ().information ( self , "Search complete" , "Solution found" )
      
   def __solve ( self , field , working_pieces , current_piece ):
      print ( "Adding piece ID:" , current_piece )
      
      if current_piece == len ( working_pieces ):
         print ( "Solution found" )
         field.print ()
         self.__puzzle_viewer.update ( field )
         
         return True
      
      working_pieces[ current_piece ].print ()
      
      print ( "Into this field:" )
      field.print ()
      self.__puzzle_viewer.update ( field )
      
      solved = False
      
      print ( "Trying to place..." )
      
      y = 0
      while ( y < field.getHeight () ) and not solved:
         x = 0
         while ( x < field.getWidth () ) and not solved:
            if field.canBeInserted ( working_pieces[ current_piece ] , x , y ):
               temp_field = copy.deepcopy ( field )
               temp_field.insertPiece ( working_pieces[ current_piece ] , x , y )
               
               print ( "Piece added at ", x, ",", y )
               
               print ( "New field:" )
               temp_field.print ()
               self.__puzzle_viewer.update ( temp_field )
               
               solved = self.__solve ( temp_field , working_pieces , current_piece + 1 )
               
               if not solved:
                  print ( "Trying another position with piece" , current_piece )
            x += 1
         y += 1
         
      return solved
